### ISSUE TITLE
User permissions for groups

### ISSUE THREAD

**Issue by jjsarton** [jech/galene#235]
I have created a group with all the possibe ppermissions.

According to the README file an user with the permissions admin may be  present.

The user with admin rigth can only raise  or unraise the hand or type /help,
/renegotiate and /leave.

For all other actions the server said: not autorized.

An Observer has more rights, it may send a private message.

I think that this is not very nice.

**Comment by jech** [comment]
That's right, the user-friendly permissions are defined here: https://github.com/jech/galene/blob/master/group/description.go#L29.  If you need more fine-grained control, you may use primitive permission sets in the config file, e.g.

    "permissions": ["admin", "op", "present", "message", "caption", "token"]

However, this is undocumented, and the exact definitions of primitive permissions might change between versions.

**Comment by jjsarton** [comment]
Okay however the admin will have (for your example) the same right as the op. "present" may be not necessary here.

A kind of user is not foreseen. A participant which has the right to enter the group with camera and microphone as the presenter but don't have the right to share and broadcast files.

**Comment by jech** [comment]
> Okay however the admin will have (for your example) the same right as the op.
> "present" may be not necessary here.

I'm not sure what you mean.

> A kind of user is not foreseen. A participant which has the right to enter the
> group with camera and microphone as the presenter but don't have the right to
> share and broadcast files.

Permissions in Galene are enforced at the server, since the client can be
modified by the user.  From the server's point of view, there's no
distinction: whether a stream comes from the camera or from an on-disk
file, it's just a video stream.  So there's no way to implement different
permissions for camera streams and broadcast files.

(Of course, the client labels all streams, so the server can know, but
there would be nothing preventing the client from lying, so it would
merely give a false sense of security.  And we try not to do false
security in Galene.)

As to sharing files, it could be implemented, but I'd need to understand
why that is needed.

**Comment by jjsarton** [comment]
> Okay however the admin will have (for your example) the same right as the op. "present" may be not necessary here.
> I'm not sure what you mean.
In your example the admin has the same right as the op so this don't make a great sense. Deleting the presenter right will preserve bandwidth (no video no sound to be received and send to all participant).
 
> A kind of user is not foreseen. A participant which has the right to enter the group with camera and microphone as the presenter but don't have the right to share and broadcast files.
> Permissions in Galene are enforced at the server, since the client can be modified by the user. From the server's point of view, there's no distinction: whether a stream comes from the camera or from an on-disk file, it's just a video stream. So there's no way to implement different permissions for camera streams and broadcast files. (Of course, the client labels all streams, so the server can know, but there would be nothing preventing the client from lying, so it would merely give a false sense of security. And we try not to do false security in Galene.) As to sharing files, it could be implemented, but I'd need to understand why that is needed.
Okay I have to deal with galene.js. I can make the allow present to the admin account and forbid within galene.js sharing and broadcasting or the better way have an other permission  within the server, for example nosharing which will be treated internally as present an within galene.js as presenter without sharing and broadcast right.

**Comment by jech** [comment]
> I can make the allow present to the admin account and forbid within galene.js sharing and broadcasting

Sure, you can do that.  However, if you do that, there's nothing preventing the user from patching the client from the JavaScript console, so it only gives a fake sense of security.

**Comment by jjsarton** [comment]
> 
> Sure, you can do that. However, if you do that, there's nothing preventing the user from patching the client from the JavaScript console, so it only gives a fake sense of security.

Nothings is perfect and patching the javascript code need some knowledge. A miss use of the admin right is also not a way which can be put within the distributed code.

**Comment by jech** [comment]
I'm not sure what you mean.

Permissions are there to enforce security.  Are you suggesting using a permission for a purpose unrelated to security?

**Comment by jjsarton** [comment]
Security is essential. I have the same meaning as you. 
If I invite people which don't have many knowledge, i am sure that they will attempt to share a screen or try to broadcast a file. This may be unwanted. 
At this time the only permission which allow to send the image from the camera and the sound from the microphone is the present permission. 
If we would have a further name which has, on the server the same purpose as present (an alias) we can check within galene.js if broadcast and screen share are permitted and delete / disable the corresponding  elements within the UI.
I have noticed, that your mean that a user may be able to inject javascript, this will not be the case for most of the user.

**Comment by jech** [comment]
So what you want is a permission that is not enforced by the server, just by the client, right?

**Comment by jjsarton** [comment]
My experience tell me that inexperienced user are able to discover problems and to try all possible things. I had with other systems as Galène no problems with screen sharing so that this may be hypothetical but will arrive.

Yes the permissions for this shall be on the client side. I have played as user with different permissions and if I type /help some commands are offered and I got, for most (observer, ...),   "The Server said: not authorised".

**Comment by jech** [comment]
>> Is that an actual problem that you have encountered, or is this purely hypothetical?

> My experience tell me that inexperienced user are able to discover problems and to try all possible things.

I'd rather not be adding features unless there's a demonstrated need for them, so I suggest we wait until this turns out to be an actual problem in practice.

> Yes the permissions for this shall be on the client side.

Then it's not a permission in the Galene sense.  If we decide it's necessary, we'll need to encode it somewhere else.

(One option would be the user.data field.)

**Comment by jech** [comment]
Closing for now, please reopen if you disagree.
